### PR TITLE
Extracting metadata from result file names using regular expressions

### PR DESCRIPTION
**Pull request by nickshulman**
This change allows users to define "Metadata Rule Sets" (any suggestions for a better name?).

Here are some screenshots:
image

image

image

image

image

### CODE REVIEW

**Comment by chambm** [comment]
What about `Annotation Rules`?  Or are the potential effects more broad than annotations?

**Comment by brendanx67** [comment]
Or Results Rules? We call these files "Results" as in File > Import > Results. Or Result File Rules. Getting a bit verbose, but "metadata" is so generic it doesn't say much and "annotation rules" doesn't tell you much either, just that the end-result it targeted at annotations.

**Comment by chambm** [comment]
Automated Annotation Extraction Rules is precise but very long. :(

**Comment by nickshulman** [comment]
Thank you for these suggestions. I think we might be heading along the right track in terms of what to call these.

The extracted values can be put into things that are not annotations, such as replicate properties such as "Sample Type", "Concentration".
You can also extract values from things other than the result file name. One example is that you could write a rule which takes the "Total Ion Current Area" value and pokes it into the "Explicit Global Standard Area" on the Result File.

In the future, I was envisioning that these rules could be used for things other than Result Files. For instance, you might want to take some values out of the protein descriptions (from a FASTA file) and set some protein annotations based on that.

**Comment by nickshulman** [comment]
At lab meeting this morning, Genn suggested maybe it should be called something involving the word "Template".
Can we think of a good name along the lines of "template", "plate layout", or "experimental design"?

File: pwiz_tools/Skyline/Model/AuditLog/PropertyNames.resx [hunk 1]
```diff
@@ -1547,9 +1547,33 @@
   <data name="IonMobilityLibrary_FilePathAuditLog" xml:space="preserve">
     <value>File path</value>
   </data>
+  <data name="DataSettings_MetadataRuleSets" xml:space="preserve">
+    <value>Metadata rule sets</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Model/AuditLog/PropertyNames.resx:1551]: Needs to change to "Result files rule sets"

File: pwiz_tools/Skyline/Model/AuditLog/PropertyNames.resx [hunk 2]
```diff
@@ -1547,9 +1547,33 @@
...
+    <value>Target</value>
+  </data>
+  <data name="MetadataRule_Name" xml:space="preserve">
+    <value>Name</value>
+  </data>
+  <data name="MetadataRule_RowSource" xml:space="preserve">
+    <value>Row source</value>
+  </data>
+  <data name="MetadataRule_Steps" xml:space="preserve">
+    <value>Steps</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Model/AuditLog/PropertyNames.resx:1575]: Needs to change to "Result files rule set" or just "Rule set". Steps is meaningless given the current UI

File: pwiz_tools/Skyline/Model/Results/ChromCacheWriter.cs [hunk 3]
```diff
@@ -174,6 +174,7 @@ protected void Complete(Exception x)
                     _completed(result, x == null ? _status : _status.ChangeErrorException(x));
                 }
                 catch (Exception x2)
+
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Model/Results/ChromCacheWriter.cs:177]: Revert this change

File: pwiz_tools/Skyline/Properties/Resources.resx [hunk 4]
```diff
@@ -11031,7 +11031,43 @@ Invalid peptide sequence at position 18
...
+    <value>Are you sure you want to delete the rule '{0}'?</value>
+  </data>
+  <data name="DocumentSettingsDlg_btnDeleteRules_Click_Are_you_sure_you_want_to_delete_these__0__rules_" xml:space="preserve">
+    <value>Are you sure you want to delete these {0} rules?</value>
+  </data>
+  <data name="DocumentSettingsDlg_ValidateMetadataRules_An_error_occurred_while_applying_the_rule___0____Do_you_want_to_continue_with_the_change_to_the_Document_Settings_" xml:space="preserve">
+    <value>An error occurred while applying the rule '{0}'. Do you want to continue with the change to the Document Settings?</value>
+  </data>
+  <data name="MetadataRuleEditor_OkDialog_There_is_already_a_metadata_rule_named___0___" xml:space="preserve">
+    <value>There is already a metadata rule named '{0}'.</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Properties/Resources.resx:11056]: Change to "result files rule set"

File: pwiz_tools/Skyline/Properties/Resources.resx [hunk 5]
```diff
@@ -11031,7 +11031,43 @@ Invalid peptide sequence at position 18
...
+    <value>Unable to find column {0}</value>
+  </data>
+  <data name="RuleError_ToString_An_error_occurred_applying_the_rule___0___" xml:space="preserve">
+    <value>An error occurred applying the rule '{0}':</value>
+  </data>
   <data name="PublishDocumentDlg_PublishDocumentDlgLoad_Go_to_Tools___Options___Panorama_tab_to_update_the_username_and_password_" xml:space="preserve">
     <value>Go to Tools &gt; Options &gt; Panorama tab to update the username and password.</value>
   </data>
+  <data name="DocumentSettingsDlg_btnDeleteRules_Click_Are_you_sure_you_want_to_delete_the_rule___0___" xml:space="preserve">
+    <value>Are you sure you want to delete the rule '{0}'?</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Properties/Resources.resx:11047]: When you refer to "rule" below, I think you mean "rule set", right? A "rule" is now what you were calling a "rule step"

File: pwiz_tools/Skyline/Properties/Resources.resx [hunk 6]
```diff
@@ -11031,7 +11031,43 @@ Invalid peptide sequence at position 18
...
+    <value>An error occurred while applying the rule '{0}'. Do you want to continue with the change to the Document Settings?</value>
+  </data>
+  <data name="MetadataRuleEditor_OkDialog_There_is_already_a_metadata_rule_named___0___" xml:space="preserve">
+    <value>There is already a metadata rule named '{0}'.</value>
+  </data>
+  <data name="MetadataRuleEditor_OkDialog__0__cannot_be_blank" xml:space="preserve">
+    <value>{0} cannot be blank</value>
+  </data>
+  <data name="MetadataRuleEditor_OkDialog_This_is_not_a_valid_regular_expression_" xml:space="preserve">
+    <value>This is not a valid regular expression.</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Properties/Resources.resx:11062]: I would prefer, "The text '{0}' is not a valid regular expression."

File: pwiz_tools/Skyline/Properties/Resources.resx [hunk 7]
```diff
@@ -11031,7 +11031,43 @@ Invalid peptide sequence at position 18
...
+  </data>
+  <data name="MetadataRuleStepEditor_OkDialog__0__must_either_be_a_valid_regular_expression_or_blank" xml:space="preserve">
+    <value>{0} must either be a valid regular expression or blank</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Properties/Resources.resx:11065]: Maybe also "The text '{0}' must..." Hate starting a sentence with something that is sure to be a mess of special characters.

File: pwiz_tools/Skyline/Properties/Resources.resx [hunk 8]
```diff
@@ -11031,7 +11031,43 @@ Invalid peptide sequence at position 18
...
+  </data>
+  <data name="MetadataRuleSetList_Title_Result_File_Rules" xml:space="preserve">
+    <value>Result File Rules</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Properties/Resources.resx:11068]: Should this be Result File Rule Set?

Comment (brendanx67) [reply pwiz_tools/Skyline/Properties/Resources.resx:11068]: Or Sets? and below would be Result File Rule Set (if it is a list of "Rules" or what were called "Rule Steps" before.

File: pwiz_tools/Skyline/SettingsUI/DefineAnnotationDlg.resx [hunk 9]
```diff
@@ -622,12 +622,12 @@
     <value>modeUIHandler</value>
   </data>
   <data name="&gt;&gt;modeUIHandler.Type" xml:space="preserve">
-    <value>pwiz.Skyline.Util.Helpers+ModeUIExtender, Skyline-daily, Version=20.1.1.8, Culture=neutral, PublicKeyToken=null</value>
+    <value>pwiz.Skyline.Util.Helpers+ModeUIExtender, Skyline-daily, Version=20.1.1.176, Culture=neutral, PublicKeyToken=null</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/SettingsUI/DefineAnnotationDlg.resx:625]: Revert this change. It is not meaningful.

File: pwiz_tools/Skyline/SettingsUI/DefineAnnotationDlg.resx [hunk 10]
```diff
@@ -622,12 +622,12 @@
...
   <data name="&gt;&gt;modeUIHandler.Type" xml:space="preserve">
-    <value>pwiz.Skyline.Util.Helpers+ModeUIExtender, Skyline-daily, Version=20.1.1.8, Culture=neutral, PublicKeyToken=null</value>
+    <value>pwiz.Skyline.Util.Helpers+ModeUIExtender, Skyline-daily, Version=20.1.1.176, Culture=neutral, PublicKeyToken=null</value>
   </data>
   <data name="&gt;&gt;$this.Name" xml:space="preserve">
     <value>DefineAnnotationDlg</value>
   </data>
   <data name="&gt;&gt;$this.Type" xml:space="preserve">
-    <value>pwiz.Skyline.Util.FormEx, Skyline-daily, Version=20.1.1.8, Culture=neutral, PublicKeyToken=null</value>
+    <value>pwiz.Skyline.Util.FormEx, Skyline-daily, Version=20.1.1.176, Culture=neutral, PublicKeyToken=null</value>
```
Comment (brendanx67) [inline pwiz_tools/Skyline/SettingsUI/DefineAnnotationDlg.resx:631]: Revert the whole file

File: pwiz_tools/Skyline/Test/ExtractedMetadataTest.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+{
+    [TestClass]
+    public class ExtractedMetadataTest : AbstractUnitTest
+    {
+        [TestMethod]
+        public void TestMetadataRuleSerialization()
+        {
+            var ruleSet = new MetadataRule("test", new []
+            {
+                new MetadataRuleStep().ChangeSource(PropertyPath.Parse("foo"))
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Test/ExtractedMetadataTest.cs:37]: I think this test needs to be expanded
